private/buf/buflsp/buflsp.go

```diff
@@ -0,0 +1,627 @@
+// Copyright 2020-2023 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package buflsp
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+	"sync"
+
+	"github.com/bufbuild/buf/private/buf/bufctl"
+	"github.com/bufbuild/buf/private/buf/bufformat"
+	"github.com/bufbuild/buf/private/buf/bufworkspace"
+	"github.com/bufbuild/buf/private/bufpkg/bufanalysis"
+	"github.com/bufbuild/buf/private/bufpkg/bufcheck/bufbreaking"
+	"github.com/bufbuild/buf/private/bufpkg/bufcheck/buflint"
+	"github.com/bufbuild/buf/private/bufpkg/bufimage"
+	"github.com/bufbuild/buf/private/bufpkg/bufmodule"
+	"github.com/bufbuild/buf/private/gen/data/datawkt"
+	"github.com/bufbuild/buf/private/pkg/app/appext"
+	"github.com/bufbuild/buf/private/pkg/ioext"
+	"github.com/bufbuild/buf/private/pkg/normalpath"
+	"github.com/bufbuild/buf/private/pkg/syserror"
+	"github.com/bufbuild/buf/private/pkg/tracing"
+	"github.com/fsnotify/fsnotify"
+	"go.lsp.dev/jsonrpc2"
+	"go.lsp.dev/protocol"
+	"go.uber.org/zap"
+)
+
+var (
+	// wellKnownTypesDescriptorProtoPath is the path of the descriptor.proto well-known types file.
+	wellKnownTypesDescriptorProtoPath = normalpath.Join("google", "protobuf", "descriptor.proto")
+	// v3WellKnownTypesCacheRelDirPath is the relative path to the cache directory for materialized
+	// well-known types .proto files.
+	v3WellKnownTypesCacheRelDirPath = normalpath.Join("v3", "wkt")
```

Comment:
I talked to @Alfus about this, but this implies it's the same cache directory as the module cache - is that correct? If so, this should be in a different location - `buf lsp` should not modify the module cache. We could have `$BUF_CACHE_DIR/lsp/wkt/WKT_VERSION` as an example

Reply:
Putting the version there as a subdirectory seems like a good idea, though I don't see an obvious version in the generated datawkt package. Digest can easily work if that's sufficient/reasonable, though.

Reply:
(For now I've addressed this by adding the digest of the wkt module into the resulting path.)